File: PythonArchive/Shortest-Peak-Reversal-5m-3x-Margin--main/Unity App Project/short_trader_multi_filter/Assets/Scripts/Core/SellOrderEngine.cs
using System;
using System.Collections.Generic;

namespace UnityApp.ShortTraderMultiFilter
{
    public class SellOrderEngine
    {
        private readonly TraderConfig _config;

        public SellOrderEngine(TraderConfig config)
        {
            _config = config;
        }

        public bool ShouldEnter(Dictionary<string, object> row, string? positionSide)
        {
            if (!string.IsNullOrEmpty(positionSide))
            {
                return false;
            }

            return row.TryGetValue("entry_signal", out var signalObj) && Convert.ToBoolean(signalObj)
                   && (!row.TryGetValue("tradable", out var tradableObj) || Convert.ToBoolean(tradableObj));
        }

        public (PositionState? Position, string Status, double EntryFee, double TradeValue, double MarginUsed) OpenPosition(
            Candle candle,
            double availableUsdt,
            bool useRawMidPrice = false)
        {
            (double? price, string status) fill;
            if (useRawMidPrice)
            {
                fill = (candle.Close, "filled");
            }
            else
            {
                fill = OrderUtils.SimulateOrderFill("short", candle.Close, _config);
            }

            if (fill.Status == "rejected" || !fill.Price.HasValue)
            {
                return (null, fill.Status, 0, 0, 0);
            }

            if (availableUsdt <= 0 || _config.RiskFraction <= 0)
            {
                return (null, "insufficient_funds", 0, 0, 0);
            }

            var marginUsed = availableUsdt * Math.Min(Math.Max(_config.RiskFraction, 0.0), _config.MaxRiskFraction);
            var leverageUsed = OrderUtils.ResolveLeverage(marginUsed, _config.DesiredLeverage, _config);
            var tradeValue = marginUsed * leverageUsed;
            if (tradeValue < _config.MinNotional)
            {
                return (null, "min_notional_not_met", 0, 0, 0);
            }

            var qty = tradeValue / fill.Price.Value;
            var entryFee = OrderUtils.BybitFee(tradeValue, _config);
            var liqPrice = OrderUtils.CalcLiquidationPriceShort(fill.Price.Value, (int)leverageUsed, _config);

            var position = new PositionState
            {
                Side = "short",
                EntryPrice = fill.Price.Value,
                LiquidationPrice = liqPrice,
                Quantity = qty,
                EntryTime = candle.Timestamp,
                EntryFee = entryFee,
                TradeValue = tradeValue,
                MarginUsed = marginUsed,
                Leverage = leverageUsed
            };
            return (position, fill.Status, entryFee, tradeValue, marginUsed);
        }
    }
}

File: PythonArchive/Shortest-Peak-Reversal-5m-3x-Margin--main/src/LIVE_short_trader_multi_filter/sell_order_engine.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Optional, Tuple

import pandas as pd

from .config import TraderConfig
from .order_utils import (
    PositionState,
    bybit_fee_fn,
    calc_liq_price_short,
    resolve_leverage,
    simulate_order_fill,
)


@dataclass
class SellOrderEngine:
    config: TraderConfig

    def should_enter(self, row: pd.Series, position_side: Optional[str]) -> bool:
        if position_side:
            return False
        return bool(row.get("entry_signal") and row.get("tradable", True))

    def open_position(
        self,
        row: pd.Series,
        available_usdt: float,
        use_raw_mid_price: bool = False,
    ) -> Tuple[Optional[PositionState], str, float, float, float, str]:
        if use_raw_mid_price:
            entry_price, status = float(row["Close"]), "filled"
        else:
            entry_price, status = simulate_order_fill("short", row["Close"], self.config)
        if status == "rejected" or entry_price is None:
            return None, status, 0.0, 0.0, 0.0, "order_rejected_or_no_fill"

        if available_usdt <= 0:
            return None, "insufficient_funds", 0.0, 0.0, 0.0, "no_balance_available"

        risk_fraction = min(max(self.config.risk_fraction, 0.0), self.config.max_risk_fraction)
        if risk_fraction == 0:
            return None, "insufficient_funds", 0.0, 0.0, 0.0, "risk_fraction_zero"

        margin_used = available_usdt * risk_fraction
        leverage_used = resolve_leverage(margin_used, self.config.desired_leverage, self.config)
        trade_value = margin_used * leverage_used
        if trade_value < self.config.min_notional:
            return None, "min_notional_not_met", 0.0, 0.0, 0.0, "below_min_notional"
        qty = trade_value / entry_price
        entry_fee = bybit_fee_fn(trade_value, self.config)

        liq_price = calc_liq_price_short(entry_price, int(leverage_used), self.config)

        position = PositionState(
            side="short",
            entry_price=entry_price,
            liq_price=liq_price,
            qty=qty,
            entry_bar_time=row.name,
            entry_fee=entry_fee,
            trade_value=trade_value,
            margin_used=margin_used,
            leverage=leverage_used,
        )
        return position, status, entry_fee, trade_value, margin_used
